### PR TITLE
feat(pdk) add documents for JavaScript and Python PDK (2.4)

### PR DESCRIPTION
**Pull request by fffonion**
### Review

@reviewer

### Testing

TODO:
copy to app/enterprise/2.4.x/external-plugins.md as well
Logged https://konghq.atlassian.net/browse/DT-1390 to port this to EE for 2.4 release.

### CODE REVIEW

**Comment by tlblessing** [comment]
@fffonion thank you for the PR. I see your comment:

> TODO: copy to app/enterprise/2.3.x/external-plugins.md as well

We can set this up as includes. I'll edit what is present, and find a spot in the TOC. The External plugins section (Go) needs some relocating as well; I have tickets for that to create a cohesive plugin PDK section.

**Comment by tlblessing** [comment]
> Also, is this a 2.4 feature rather than 2.3? That was my impression from @PaulFischer-Kong from a meeting we had recently.

Paul has confirmed this is a 2.4.x feature, so this PR needs to be retargeted:

> hi! :wave: yes that is correct
> 
> 2.4 feature

**Comment by fffonion** [comment]
@tlblessing This is indeed a 2.4 feature, but specially the feature doesn't occur inside Kong itself, so it's also supported on  Kong Gateway/Enterprise 2.3. So I'm thinking backporting it to 2.3 docs is also helpful. How do you think?

File: app/gateway-oss/2.3.x/external-plugins.md [hunk 1]
```diff
@@ -242,6 +242,216 @@ pluginserver_my_plugin_query_cmd = /usr/local/bin/my-plugin -dump
...
 
+## Developing JavaScript plugins
+
+{{site.base_gateway}} support for the JavaScript language is provided by [kong-js-pdk].
+The library provides a plugin server to provide runtime for JavaScript plugins, and
+functions to access {{site.base_gateway}} features of the [PDK][kong-pdk].
+
+TypeScript is also supported in the following ways:
+
+- [kong-js-pdk] includes type definitions for PDK functions that allows type checking
```
Comment (tlblessing) [inline app/gateway-oss/2.3.x/external-plugins.md:253]:
```suggestion
- [kong-js-pdk] includes type definitions for PDK functions that allow type checking
```

File: app/gateway-oss/2.3.x/external-plugins.md [hunk 2]
```diff
@@ -242,6 +242,216 @@ pluginserver_my_plugin_query_cmd = /usr/local/bin/my-plugin -dump
...
+when developing plugins in TypeScript.
+- Plugin written in TypeScript can be loaded directly and transpiled on-the-fly.
```
Comment (tlblessing) [inline app/gateway-oss/2.3.x/external-plugins.md:255]:
```suggestion
- Plugin written in TypeScript can be loaded directly and transpiled on the fly.
```

File: app/gateway-oss/2.3.x/external-plugins.md [hunk 3]
```diff
@@ -242,6 +242,216 @@ pluginserver_my_plugin_query_cmd = /usr/local/bin/my-plugin -dump
...
+
+### Development
+
+[kong-js-pdk] can be installed via `npm`. To install the plugin server binary globally, use:
```
Comment (tlblessing) [inline app/gateway-oss/2.3.x/external-plugins.md:259]:
```suggestion
[kong-js-pdk] can be installed using `npm`. To install the plugin server binary globally:
```

File: app/gateway-oss/2.3.x/external-plugins.md [hunk 4]
```diff
@@ -242,6 +242,216 @@ pluginserver_my_plugin_query_cmd = /usr/local/bin/my-plugin -dump
...
+
+```
+npm install kong-pdk -g
+```
+
+To use it in your local development directory, use:
```
Comment (tlblessing) [inline app/gateway-oss/2.3.x/external-plugins.md:265]:
```suggestion
To install it in your local development directory:
```

File: app/gateway-oss/2.3.x/external-plugins.md [hunk 5]
```diff
@@ -242,6 +242,216 @@ pluginserver_my_plugin_query_cmd = /usr/local/bin/my-plugin -dump
...
+
+```
+npm install kong-pdk --save
+```
+
+A valid JavaScript plugin implementation should export following object:
```
Comment (tlblessing) [inline app/gateway-oss/2.3.x/external-plugins.md:271]:
```suggestion
A valid JavaScript plugin implementation should export the following object:
```

File: app/gateway-oss/2.3.x/external-plugins.md [hunk 6]
```diff
@@ -242,6 +242,216 @@ pluginserver_my_plugin_query_cmd = /usr/local/bin/my-plugin -dump
...
+- `response`
+- `preread`
+- `log`
+
+Similar to Lua plugins, the presence of the `response` handler automatically enables the buffered proxy mode.
+
+#### PDK functions
+
+[kong-js-pdk] invokes PDK functions in Kong through network-based RPC. Thus each function returns a Promise
+instance; it's convient to use `async`/`await` keywords in phase handlers for better readibility.
```
Comment (tlblessing) [inline app/gateway-oss/2.3.x/external-plugins.md:323]:
```suggestion
instance; it's convenient to use `async`/`await` keywords in phase handlers for better readability.
```

**Review by tlblessing — CHANGES_REQUESTED** [review]
Thank you for the PR @fffonion. When we have a 2.4.x branch, we can retarget this PR, and I can set up includes to use between OSS and Enterprise. Update: Per @lena-larionova, no includes at this time. 

I did an initial review for now with suggested minor edits for typos.

**Comment by tlblessing** [comment]
> @tlblessing This is indeed a 2.4 feature, but specially the feature doesn't occur inside Kong itself, so it's also supported on Kong Gateway/Enterprise 2.3. So I'm thinking backporting it to 2.3 docs is also helpful. How do you think?

@fffonion I'm not sure since @PaulFischer-Kong siad above this is 2.4 feature. We do have docs for external plugins as you know, and this PR needs to 

> @tlblessing This is indeed a 2.4 feature, but specially the feature doesn't occur inside Kong itself, so it's also supported on Kong Gateway/Enterprise 2.3. So I'm thinking backporting it to 2.3 docs is also helpful. How do you think?

@fffonion I'm not sure since @PaulFischer-Kong said above this is a 2.4 feature?

**Review by kikito — APPROVED** [review]

**Comment by tlblessing** [comment]
@fffonion there is another PR by @javierguerragiraldez updating this file for Python PDK in PR https://github.com/Kong/docs.konghq.com/pull/2741.

**Comment by tlblessing** [comment]
Direct review link:

https://deploy-preview-2690--kongdocs.netlify.app/gateway-oss/2.4.x/external-plugins/

(Had to update redirects)

**Review by tlblessing — APPROVED** [review]

**Pull request merged**